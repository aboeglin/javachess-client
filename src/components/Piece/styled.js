import styled from "styled-components";

export const Container = styled.div`
  height: 100%;

  img {
    cursor: pointer;
    height: 44px;
    position: absolute;
    display: block;
    bottom: 18px;
    left: 50%;
    transform: translateX(-50%);
  }
`;
