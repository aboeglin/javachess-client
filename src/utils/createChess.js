import SockJS from "sockjs-client";
import Stomp from "stompjs";

import { pipe, prop, append, forEach } from "ramda";

const HARDCODED_EMAIL = "user@domain.tld";

const createChess = () => {
  const socket = new SockJS("http://localhost:8080/ws");
  const ws = Stomp.over(socket);

  ws.debug = null;

  let observers = [];
  let gameId = null;

  let state = {
    pieces: [],
    possibleMoves: [],
    selection: null
  };

  ws.connect({}, () => {
    ws.subscribe("/user/queue/lfg/ack", frame => {
      gameId = pipe(prop("body"), JSON.parse, prop("gameId"))(frame);
      ws.subscribe(
        `/queue/game/${gameId}/ready`,
        pipe(prop("body"), body => {
          const newPieces = JSON.parse(body).game.board.pieces;
          updateState({ pieces: newPieces });
        })
      );

      ws.subscribe(
        `/queue/game/${gameId}/piece-moved`,
        pipe(prop("body"), body => {
          const newPieces = JSON.parse(body).game.board.pieces;

          // If move successful
          updateState({
            pieces: newPieces,
            possibleMoves: [],
            selection: null
          });
          // Else API should return MoveError and we should keep the possible moves
        })
      );

      ws.subscribe(
        `/queue/game/${gameId}/possible-moves`,
        pipe(prop("body"), body => {
          const possibleMoves = JSON.parse(body).possibleMoves;
          updateState({ possibleMoves: possibleMoves });
        })
      );

      ws.send(
        `/app/game/${gameId}/join`,
        {},
        JSON.stringify({ email: HARDCODED_EMAIL })
      );
    });

    ws.subscribe("/user/queue/errors", frame => {
      console.error(frame.body);
    });

    ws.send("/app/lfg", {}, JSON.stringify({ email: HARDCODED_EMAIL }));
  });

  const pieceSelected = piece => {
    updateState({ selection: piece });
    ws.send(
      `/app/game/${gameId}/select-piece`,
      {},
      JSON.stringify({ email: HARDCODED_EMAIL, x: piece.x, y: piece.y })
    );
  };

  const pieceMoved = (fromX, fromY, toX, toY) => {
    ws.send(
      `/app/game/${gameId}/perform-move`,
      {},
      JSON.stringify({ email: HARDCODED_EMAIL, fromX, fromY, toX, toY })
    );
  };

  const updateState = newState => {
    state = { ...state, ...newState };
    forEach(fn => fn(state))(observers);
  };

  // Store array of handlers, otherwise only the last one can get notified from state changed
  const onStateChanged = fn => {
    observers = append(fn, observers);
    fn(state);
  };

  return {
    onStateChanged,
    pieceSelected,
    pieceMoved
  };
};

export default createChess;
