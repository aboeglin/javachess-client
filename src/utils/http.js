import F from "fluture";
import axios from "axios";
import { curry } from "ramda";

export const post = curry((url, body) =>
  F((bad, good) => {
    axios
      .post(url, body)
      .then(good)
      .catch(bad);

    // Cancelation function
    return () => {};
  })
);

export const patch = curry((url, body) =>
  F((bad, good) => {
    axios
      .patch(url, body)
      .then(good)
      .catch(bad);

    // Cancelation function
    return () => {};
  })
);

export const get = (url) =>
  F((bad, good) => {
    axios
      .get(url)
      .then(good)
      .catch(bad);

    // Cancelation function
    return () => {};
  });
