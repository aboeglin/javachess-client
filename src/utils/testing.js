import React from "react";
import ReactDOM from "react-dom";
import { keys, pipe, propOr, curry } from "ramda";
import renderer from "react-test-renderer";
import cleanStack from "clean-stack";
import { Helmet } from "react-helmet";

beforeEach(() => {
  process.env.GATSBY_ENTRYPOINT = "/";
});

export const hasNonEmptyPropTypes = x => {
  const propTypes = propOr(false, "propTypes", x);
  // () => <span>cool</span> is valid, so no need for propTypes
  if (!propTypes) return true;
  // expect some keys if defined
  return pipe(keys, z => z.length > 0)(propTypes);
};

export const render = (C, props = {}, el = "div") => {
  try {
    const div = document.createElement(el);
    ReactDOM.render(
      <>
        <div id="leafletmap" />
        <C {...props} />
      </>,
      div
    );
    ReactDOM.unmountComponentAtNode(div);
  } catch (error) {
    console.error(error.stack ? cleanStack(error.stack) : error);
  }
};
export const renderize = (C, props = {}, el = "div") => () => {
  try {
    const rend = () => {
      try {
        render(C, props, el);
      } catch (error) {
        if (error.stack) error.stack = cleanStack(error.stack);
        console.error(error.stack ? error.stack : error);
        throw error;
      }
    };
    expect(rend).not.toThrow();
  } catch (error) {
    console.error(error.stack ? cleanStack(error.stack) : error);
  }
};

const nameOf = C => C.displayName || C.name || "A Component";

export const renderWithTestRenderer = curry((C, props) => {
  it(`renders <${nameOf(C)} />`, () => {
    try {
      const c = renderer.create(<C {...props} />);
      let tree = c.toJSON();
      expect(tree).toMatchSnapshot();
    } catch (error) {
      console.log(error.stack ? cleanStack(error.stack) : error);
    }
  });
});

export const standardComponentTest = (C, props = {}, el = "div") => {
  renderWithTestRenderer(C, props);
  it("renders without crashing", renderize(C, props, el));
  it("has propTypes", () => {
    expect(hasNonEmptyPropTypes(C)).toBeTruthy();
  });
};

export const standardComponentTestWithHeadTest = (
  C,
  props = {},
  el = "div",
  headProperty = "",
  headData = null
) => {
  renderWithTestRenderer(C, props);
  it("renders without crashing", renderize(C, props, el));
  it("has propTypes", () => {
    expect(hasNonEmptyPropTypes(C)).toBeTruthy();
  });
  it("has correct head props", () => {
    headData
      ? expect(Helmet.peek()).toHaveProperty(headProperty, headData)
      : expect(Helmet.peek()).toHaveProperty(headProperty);
  });
};
