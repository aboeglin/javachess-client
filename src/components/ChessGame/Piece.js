import React from "react";

import { curry, pipe } from "ramda";

import { PieceContainer } from "./styled";

import BLACK_PAWN_IMAGE_URL from "./images/pawn-black.png";
import BLACK_BISHOP_IMAGE_URL from "./images/bishop-black.png";
import BLACK_KNIGHT_IMAGE_URL from "./images/knight-black.png";
import BLACK_ROOK_IMAGE_URL from "./images/rook-black.png";
import BLACK_QUEEN_IMAGE_URL from "./images/queen-black.png";
import BLACK_KING_IMAGE_URL from "./images/king-black.png";
import WHITE_PAWN_IMAGE_URL from "./images/pawn-white.png";
import WHITE_BISHOP_IMAGE_URL from "./images/bishop-white.png";
import WHITE_KNIGHT_IMAGE_URL from "./images/knight-white.png";
import WHITE_ROOK_IMAGE_URL from "./images/rook-white.png";
import WHITE_QUEEN_IMAGE_URL from "./images/queen-white.png";
import WHITE_KING_IMAGE_URL from "./images/king-white.png";
import withChess from "@hocs/withChess";

const getImageUrl = (type, color) =>
  ({
    BLACK_PAWN: BLACK_PAWN_IMAGE_URL,
    BLACK_BISHOP: BLACK_BISHOP_IMAGE_URL,
    BLACK_KNIGHT: BLACK_KNIGHT_IMAGE_URL,
    BLACK_ROOK: BLACK_ROOK_IMAGE_URL,
    BLACK_QUEEN: BLACK_QUEEN_IMAGE_URL,
    BLACK_KING: BLACK_KING_IMAGE_URL,
    WHITE_PAWN: WHITE_PAWN_IMAGE_URL,
    WHITE_BISHOP: WHITE_BISHOP_IMAGE_URL,
    WHITE_KNIGHT: WHITE_KNIGHT_IMAGE_URL,
    WHITE_ROOK: WHITE_ROOK_IMAGE_URL,
    WHITE_QUEEN: WHITE_QUEEN_IMAGE_URL,
    WHITE_KING: WHITE_KING_IMAGE_URL
  }[`${color}_${type}`]);

const handlePieceClicked = curry(
  (piece, playerColor, activePlayerColor, pieceSelected) => () => {
    if (piece.color === playerColor && activePlayerColor === playerColor) {
      pieceSelected(piece);
    }
  }
);

export const Piece = ({
  piece,
  pieceSelected,
  playerColor,
  activePlayerColor
}) =>
  pipe(
    ({ type, color }) => getImageUrl(type, color),
    url => (
      <PieceContainer
        type={piece.type}
        x={piece.x}
        onClick={handlePieceClicked(
          piece,
          playerColor,
          activePlayerColor,
          pieceSelected
        )}
      >
        <img src={url} />
      </PieceContainer>
    )
  )(piece);

export default withChess(Piece);
