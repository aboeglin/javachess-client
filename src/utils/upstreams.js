import { __, concat, curry, map, pipe, prop, toString } from "ramda";

import { get, patch, post } from "./http";

const ENDPOINT =
  process.env.GATSBY_CHESS_SERVICE_ENDPOINT || "http://localhost:8080";
const GET_GAMES_PATH = "/games";
const POST_GAMES_PATH = "/games";
const PATCH_GAMES_PATH = "/games";

export const fetchGamesWithGetAndEndpoint = curry((get, endpoint, path) =>
  pipe(concat(endpoint), get, map(prop("data")))(path)
);

export const fetchGames = () =>
  fetchGamesWithGetAndEndpoint(get, ENDPOINT, GET_GAMES_PATH);

export const createGameWithPostAndEndpoint = curry(
  (post, endpoint, path, userName) =>
    pipe(
      concat(endpoint),
      post(__, { userId: userName }),
      map(prop("data"))
    )(path)
);

export const createGame = createGameWithPostAndEndpoint(
  post,
  ENDPOINT,
  POST_GAMES_PATH
);

export const joinGameWithPatchAndEndpoint = curry(
  (patch, endpoint, path, userName, gameId) =>
    pipe(
			toString,
      concat("/"),
      concat(path),
      concat(endpoint),
      patch(__, { userId: userName }),
      map(prop("data"))
    )(gameId)
);

export const joinGame = joinGameWithPatchAndEndpoint(patch, ENDPOINT, PATCH_GAMES_PATH);
