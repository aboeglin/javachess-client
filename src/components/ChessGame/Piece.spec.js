import { Piece } from "./Piece";
import { standardComponentTest } from "../../utils/testing";

standardComponentTest(Piece, {
  piece: { type: "PAWN", color: "WHITE", x: "a", y: "4" },
  pieceSelected: { type: "PAWN", color: "WHITE", x: "b", y: "4" }
});
