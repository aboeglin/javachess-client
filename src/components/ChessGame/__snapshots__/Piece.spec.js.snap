// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders <Piece /> 1`] = `
<div
  className="css-132xnrr"
  onClick={[Function]}
  type="PAWN"
  x="a"
>
  <img
    src={Object {}}
  />
</div>
`;
